### PR TITLE
`rustc_parse` top-level cleanups

### PR DESCRIPTION
**Pull request by nnethercote**
A bunch of improvements in and around `compiler/rustc_parse/src/lib.rs`. Many of the changes streamline the API in that file from this (12 functions and one macro):
```
    name                              args                  return type
    ----                              ----                  -----------
    panictry_buffer!                  Result<T, Vec<Diag>>  T

pub parse_crate_from_file             path                  PResult<Crate>
pub parse_crate_attrs_from_file       path                  PResult<AttrVec>
pub parse_crate_from_source_str       name,src              PResult<Crate>
pub parse_crate_attrs_from_source_str name,src              PResult<AttrVec>

pub new_parser_from_source_str        name,src              Parser
pub maybe_new_parser_from_source_str  name,src              Result<Parser, Vec<Diag>>
pub new_parser_from_file              path,error_sp         Parser
    maybe_source_file_to_parser       srcfile               Result<Parser, Vec<Diag>>

pub parse_stream_from_source_str      name,src,override_sp  TokenStream
pub source_file_to_stream             srcfile,override_sp   TokenStream
    maybe_file_to_stream              srcfile,override_sp   Result<TokenStream, Vec<Diag>>

pub stream_to_parser                  stream,subparser_name Parser
```
to this:
```
    name                              args                  return type
    ----                              ----                  -----------
    unwrap_or_emit_fatal              Result<T, Vec<Diag>>  T
  
pub new_parser_from_source_str        name,src              Result<Parser, Vec<Diag>>
pub new_parser_from_file              path,error_sp         Result<Parser, Vec<Diag>>
    new_parser_from_source_file       srcfile               Result<Parser, Vec<Diag>>

pub source_str_to_stream              name,src,override_sp  Result<TokenStream, Vec<Diag>>
    source_file_to_stream             srcfile,override_sp   Result<TokenStream, Vec<Diag>>
```
I found the old API quite confusing, with lots of similar-sounding function names and no clear structure. I think the new API is much better.

r? @spastorino

### CODE REVIEW

**Comment by nnethercote** [comment]
Best reviewed one commit at a time. The commit messages explain each commit.

File: src/tools/rustfmt/src/parse/parser.rs [hunk 1]
```diff
@@ -68,10 +69,10 @@ impl<'a> ParserBuilder<'a> {
     ) -> Result<rustc_parse::parser::Parser<'a>, Option<Vec<Diag<'a>>>> {
         match input {
             Input::File(ref file) => catch_unwind(AssertUnwindSafe(move || {
-                new_parser_from_file(psess, file, None)
+                unwrap_or_emit_fatal(new_parser_from_file(psess, file, None))
```
[the remark this answers is not in the document]
Comment (nnethercote) [reply src/tools/rustfmt/src/parse/parser.rs:72]: Correct. `unwrap_or_emit_fatal(new_parser_from_file(...))` is equivalent to the old `new_parser_from_file(...)`. So there is no change to existing behaviour.

A nice thing about this is that the `catch_unwind` will be removable in a follow-up, removing the asymmetry between the `Input::File` and `Input::Text` cases here.

[the remark this answers is not in the document]
Comment (ytmimi) [reply src/tools/rustfmt/src/parse/parser.rs:72]: That's great! I appreciate the explanation.

[the remark this answers is not in the document]
Comment (nnethercote) [reply src/tools/rustfmt/src/parse/parser.rs]: I've had time to do the follow-up, which I added as an extra commit to this PR.

[the remark this answers is not in the document]
Comment (ytmimi) [reply src/tools/rustfmt/src/parse/parser.rs]: Awesome! Thanks for the updated :)

**Comment by nnethercote** [comment]
I rebased.

@bors r=spastorino

**Pull request merged**